cpu/stm32/periph/fdcan.c

```diff
@@ -0,0 +1,1737 @@
+
+    /* Enable auto retransmission on failure*/
+    can->CCCR &= ~FDCAN_CCCR_DAR;
+    /* Enable FD mode */
+    can->CCCR |= FDCAN_CCCR_FDOE;
+    /* Enable bitrate switching */
+    can->CCCR |= FDCAN_CCCR_BRSE;
+    /* Enable transmit pause */
+    can->CCCR |= FDCAN_CCCR_TXP;
+    /* Enable protocol exception handling */
+    can->CCCR &= ~FDCAN_CCCR_PXHD;
+
+#ifdef STM32_PM_STOP
+    pm_block(STM32_PM_STOP);
+#endif
+
+    /* Clear interrupt flags */
+    can->IR |= 0xFFFFFFU;
+    /* Enable all interrupts and both interrupt lines */
+    can->IE |= 0xFFFFFFU;
+
+    can->TXBTIE = FDCAN_TXBTIE_TIE;
+
+    can->ILS = FDCAN_ILS_PERR | FDCAN_ILS_SMSG;
+    can->ILE |= FDCAN_ILE_EINT0 | FDCAN_ILE_EINT1;
+
+    DEBUG("%s: FDCAN%u->CCCR =  %lx\n",
+          __func__, get_channel_id(can), can->CCCR);
+
+    int res = set_mode(can, MODE_NORMAL);
+
+    DEBUG("%s: FDCAN%u init done, return %d\n",
+          __func__, get_channel_id(can), res);
+
+    return res;
+}
+
+static inline void set_bit_timing(can_t *dev)
+{
+    FDCAN_GlobalTypeDef *can = dev->conf->can;
+
+    DEBUG("%s: FDCAN%u setup bittiming\n",
+          __func__, get_channel_id(can));
+    /* Normal CAN bittiming */
+    /* SJW */
+    can->NBTP &= ~FDCAN_NBTP_NSJW;
+    can->NBTP |= (dev->candev.bittiming.sjw - 1) << FDCAN_NBTP_NSJW_Pos;
+    /* Phase Seg 2 */
+    can->NBTP &= ~FDCAN_NBTP_NTSEG2;
+    can->NBTP |= (dev->candev.bittiming.phase_seg2 - 1) << FDCAN_NBTP_NTSEG2_Pos;
+    /* Phase Seg 1 */
+    can->NBTP &= ~FDCAN_NBTP_NTSEG1;
+    can->NBTP |= (dev->candev.bittiming.phase_seg1
+                 + dev->candev.bittiming.prop_seg - 1) << FDCAN_NBTP_NTSEG1_Pos;
+    /* BRP */
+    can->NBTP &= ~FDCAN_NBTP_NBRP;
+    can->NBTP |= (dev->candev.bittiming.brp - 1) << FDCAN_NBTP_NBRP_Pos;
+
+    DEBUG("%s: FDCAN%u->NBTP = %lx\n",
+          __func__, get_channel_id(can), can->NBTP);
+
+    /* FD CAN bittiming */
+    /* SJW */
+    can->DBTP &= ~FDCAN_DBTP_DSJW;
+    can->DBTP |= (dev->candev.fd_data_bittiming.sjw - 1) << FDCAN_DBTP_DSJW_Pos;
+    /* Phase Seg 2 */
+    can->DBTP &= ~FDCAN_DBTP_DTSEG2;
+    can->DBTP |= (dev->candev.fd_data_bittiming.phase_seg2 - 1) << FDCAN_DBTP_DTSEG2_Pos;
+    /* Phase Seg 1 */
+    can->DBTP &= ~FDCAN_DBTP_DTSEG1;
+    can->DBTP |= (dev->candev.fd_data_bittiming.phase_seg1
+                 + dev->candev.fd_data_bittiming.prop_seg - 1) << FDCAN_DBTP_DTSEG1_Pos;
+    /* BRP */
+    can->DBTP &= ~FDCAN_DBTP_DBRP;
+    can->DBTP |= (dev->candev.fd_data_bittiming.brp - 1) << FDCAN_DBTP_DBRP_Pos;
+    /* Transceiver delay compensation */
+    can->DBTP |= FDCAN_DBTP_TDC;
+
+    DEBUG("%s: FDCAN%u->DBTP = %lx\n",
+          __func__, get_channel_id(can), can->DBTP);
+
+    can->TDCR &= ~FDCAN_TDCR_TDCO;
+    can->TDCR |= (dev->candev.loop_delay / dev->candev.bittiming.tq) << FDCAN_TDCR_TDCO_Pos;
+
+    DEBUG("%s: FDCAN%u->TDCR = %lx\n",
+          __func__, get_channel_id(can), can->TDCR);
+}
+
+static uint32_t fddlc_to_dlc(__uint32_t fddlc) {
```

Comment:
if dlc is wrong in the struct can_frame it is even more wrong in this function name, i mean it is not dlc that is wrong but fddlc since that is len while dlc is dlc the dlc of a fd frame does not represent its len

Reply:
> if dlc is wrong in the struct can_frame it is even more wrong in this function name, i mean it is not dlc that is wrong but fddlc since that is len while dlc is dlc the dlc of a fd frame does not represent its len

Completly right, I noticed that after the last push, already fixed it but not pushed yet, I will do it with other fixes.

Reply:
done